### PR TITLE
Add React Support

### PR DESCRIPTION
**Pull request by zurbergram**
## Description
https://github.com/department-of-veterans-affairs/vets.gov-team/issues/19826

A a developer I need to add the React library to GIDS so that it has the ability to use the same tech stack as the Comparison Tool.

Done following instructions at https://github.com/reactjs/react-rails#rails-5x
For issue that caused most of the builds failures, fix was https://github.com/rails/webpacker/issues/522#issuecomment-310006338
which created `bin/yarn`

## Testing done
- local testing

## Acceptance criteria
- [ ] The React library is included as part of the GIDS project.
- [ ] Smoke testing passes

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Review by shawkey-daniel — APPROVED** [review]
LGTM

**Comment by zurbergram** [comment]
This PR contains several fixes to tests unrelated to changes, most of these tests were using https://apidock.com/rails/v5.2.3/ActiveRecord/FinderMethods/first instead of https://apidock.com/rails/v5.2.3/ActiveRecord/FinderMethods/find_by

**Review by sonntag-adam — APPROVED** [review]

**Comment by zurbergram** [comment]
Starting around https://github.com/department-of-veterans-affairs/gibct-data-service/pull/471/commits/b1b51d75a4aad0bd3e04860c2de7c003bf851156 commits were around trying to resolve an issue that was easily reproducable in the Jenkins build but not as easy to reproduce locally

The final solution was https://github.com/rails/webpacker/issues/522#issuecomment-310006338
which created `bin/yarn`

File: Dockerfile [hunk 1]
```diff
@@ -21,6 +21,13 @@ RUN apt-get install -y libfreetype6 libfreetype6-dev libfontconfig1 libfontconfi
 RUN curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > /cc-test-reporter
 RUN chmod +x /cc-test-reporter
 
+ENV YARN_VERSION 1.12.3
+ENV NODE_ENV production
+
+RUN curl -sL https://deb.nodesource.com/setup_10.x | bash - && \
```
Comment (jholton) [inline Dockerfile]: FYI, I think we're going to need to consult with devops to get your desired versions of `node` and `yarn` in deployed environments.  Looks like versions are currently derived from [this file](https://github.com/department-of-veterans-affairs/devops/blob/master/ansible/build/roles/common/nodejs/vars/main.yml).

Comment (zurbergram) [reply Dockerfile]: https://github.com/department-of-veterans-affairs/vets-website/blob/master/Dockerfile#L9 has yarn version as 1.12.3 and is using a `node:10` image as it's base image

I'm not seeing `YARN_VERSION` listed as a property in https://github.com/department-of-veterans-affairs/vets-website/blob/master/docker-compose.yml which would enable the changing of the value I believe

File: Makefile [hunk 2]
```diff
@@ -13,7 +13,11 @@ bash:
 
 .PHONY: ci
 ci:
-	@$(BASH_TEST) "bin/rails db:setup db:migrate ci"
+	@$(BASH_TEST) "bin/rails db:setup db:migrate assets:precompile ci"
+
+.PHONY: ci_local
+ci_local:
+	@$(BASH_TEST) "RAILS_ENV=test CI=true bin/rails db:create db:schema:load db:migrate assets:precompile ci"
```
Comment (jholton) [inline Makefile]: I'm not a Make expert, but it looks like this works since the only thing different is the `CI` env variable:
```suggestion
	$(MAKE) ci CI=true"
```

**Comment by omgitsbillryan** [comment]
> A a developer I need to add the React library to GIDS so that it has the ability to use the same tech stack as the Comparison Tool.

Can you elaborate a bit more on this? My experience with GIDS is fairly limited.

**Comment by zurbergram** [comment]
@department-of-veterans-affairs/vsp-operations Could I get a code review on this at your earliest convenience. Thank you!

Specifically looking for feedback about node and yarn in deployed environments to ensure these changes will not cause issues in dev/staging environments.

File: Makefile (hunk 2 above)
Comment (zurbergram) [reply Makefile]: changed

**Comment by zurbergram** [comment]
> > A a developer I need to add the React library to GIDS so that it has the ability to use the same tech stack as the Comparison Tool.
> 
> Can you elaborate a bit more on this? My experience with GIDS is fairly limited.

The Comparison Tool in reference is https://www.va.gov/gi-bill-comparison-tool/ . 

The idea behind adding React to this repo is that over time the UI can be written with React for easier maintainability.  This is so that when switching from development in https://github.com/department-of-veterans-affairs/vets-website/ to development in https://github.com/department-of-veterans-affairs/gibct-data-service/ there is less of skill shift, additionally there is functionality that React provides that without access to require "creative" solutions to be devised to achieve desired behavior in [GIDS UI](https://staging.va.gov/gids/)

File: config/application.rb [hunk 3]
```diff
@@ -42,5 +42,19 @@ class Application < Rails::Application
...
+    config.react.server_renderer_timeout    ||= 20 # seconds
+    config.react.server_renderer = React::ServerRendering::BundleRenderer
+    config.react.server_renderer_options = {
+        files: ["server_rendering.js"],       # files to load for prerendering
+        replay_console: true,                 # if true, console.* will be replayed client-side
+    }
+    # Changing files matching these dirs/exts will cause the server renderer to reload:
+    config.react.server_renderer_extensions = ["jsx", "js"]
+    config.react.server_renderer_directories = ["/app/assets/javascripts", "/app/javascript/"]
+    config.react.camelize_props = true # default false
```
Comment (wyattwalter) [inline config/application.rb]: Does this block of configuration mean that Rails is going to serve up the React bits as well? No additional nodejs server required?

File: Dockerfile (hunk 1 above)
Comment (wyattwalter) [reply Dockerfile]: @jholton is right about versions.. but that's the default for the module. We can specify newer versions, just need to keep it in sync with what's here.

Comment (wyattwalter) [reply Dockerfile]: So this is going to just keep updating to 10. latest right? It looks like it's being pinned to 10.15.3 for the review instances on vets-website https://github.com/department-of-veterans-affairs/devops/blob/master/ansible/roles/review-instance-configure/meta/main.yml

Comment (zurbergram) [reply Dockerfile]: Updated Dockerfile to install a specific version of node using nvm, currently set to 10.15.3 via NODEJS_VERSION environment variable

File: Makefile [hunk 4]
```diff
@@ -13,7 +14,11 @@ bash:
 
 .PHONY: ci
 ci:
-	@$(BASH_TEST) "bin/rails db:setup db:migrate ci"
+	@$(BASH_TEST) "bin/rails db:setup db:migrate assets:precompile ci"
```
Comment (omgitsbillryan) [inline Makefile]: Why is it necessary to call `assets:precompile` before the `ci` rake task?

File: config/application.rb (hunk 3 above)
Comment (zurbergram) [reply config/application.rb]: These are the default configurations values when using https://github.com/reactjs/react-rails#server-side-rendering

I added them to here even though they are defaults for awareness of configuration that can change

File: Makefile [hunk 5]
```diff
@@ -13,7 +14,11 @@ bash:
...
+
+.PHONY: ci_local
+ci_local:
+	@$(MAKE) "CI=true RAILS_ENV=test ci"
```
Comment (omgitsbillryan) [inline Makefile]: I'm behind on my `makefile` syntax - is this force setting `CI=true RAILS_ENV=test` before calling `make ci`? I think `CI=true` means that this is the CI run, which for us means Jenkins. I would expect this to _not_ be set when running locally.

File: spec/models/institution_builder_spec.rb [hunk 6]
```diff
@@ -552,20 +552,20 @@
...
 
           it 'prefers Sec702School over Section702' do
             create :weam, :institution_builder, :private
-            create :sec702_school, :institution_builder, sec_702: true
+            sec702_school = create :sec702_school, :institution_builder, sec_702: true
             create :sec702, :institution_builder
             described_class.run(user)
 
-            expect(institutions.first.sec_702).to be_truthy
+            expect(institutions.find_by(facility_code: sec702_school.facility_code).sec_702).to be_truthy
```
Comment (omgitsbillryan) [inline spec/models/institution_builder_spec.rb]: I don't see how these spec changes are related to `webpacker` & `react`.

**Review by omgitsbillryan — COMMENTED** [review]
Mostly questions. 

I'm finding this PR hard to follow. Probably because I'm not a `webpacker` nor `react` expert. Is there any way to reduce the PR size?

File: Makefile (hunk 4 above)
Comment (zurbergram) [reply Makefile]: `assets:precompile` causes `webpacker:compile` to run which is needed to create `public/packs-test/manifest.json`

`public/packs-test` is ignored in the `.gitignore` which causes `public/packs-test/manifest.json` to not exist during test run which creates error below which can be seen in full at http://jenkins.vfs.va.gov/blue/organizations/jenkins/testing%2Fgibct-data-service/detail/react/13/pipeline

```
  3) DashboardsController behaves like an authenticating controller when being visited and submitting credentials outputs an error message if not successful

     Failure/Error: <%= javascript_pack_tag 'application' %>

     ActionView::Template::Error:
       Webpacker can't find application in /src/gibct/public/packs-test/manifest.json. Possible causes:
       1. You want to set webpacker.yml value of compile to true for your environment
          unless you are using the `webpack -w` or the webpack-dev-server.
       2. webpack has not yet re-run to reflect updates.
       3. You have misconfigured Webpacker's config/webpacker.yml file.
       4. Your webpack configuration is not creating a manifest.
       Your manifest contains:
       {
       }
     .
     .
     .
     .
     .
     .
     # --- Caused by: ---
     # Webpacker::Manifest::MissingEntryError:
     #   Webpacker can't find application in /src/gibct/public/packs-test/manifest.json. Possible causes:
     #   1. You want to set webpacker.yml value of compile to true for your environment
     #      unless you are using the `webpack -w` or the webpack-dev-server.
     #   2. webpack has not yet re-run to reflect updates.
     #   3. You have misconfigured Webpacker's config/webpacker.yml file.
     #   4. Your webpack configuration is not creating a manifest.
     #   Your manifest contains:
     #   {
     #   }
     #   /usr/local/bundle/gems/webpacker-4.0.7/lib/webpacker/manifest.rb:79:in `handle_missing_entry'
```

File: Makefile (hunk 5 above)
Comment (zurbergram) [reply Makefile]: running just `make ci` locally fails with error messages that are resolved by setting `CI=true` and `RAILS_ENV=test`

I can remove this addition if as it was mostly for ease of use locally

File: spec/models/institution_builder_spec.rb (hunk 6 above)
Comment (zurbergram) [reply spec/models/institution_builder_spec.rb]: Most of these tests were using https://apidock.com/rails/v5.2.3/ActiveRecord/FinderMethods/first instead of https://apidock.com/rails/v5.2.3/ActiveRecord/FinderMethods/find_by

which is an issue when you can't guarantee a clean test database for each run, additionally the assumption by a unit test shouldn't be made that the record is the only record in the database.  There have been other tests in past week or so that failed for same reason

Comment (omgitsbillryan) [reply spec/models/institution_builder_spec.rb]: OK. Sounds like it's not at all related to `webpacker` and/or `react-rails`. Should probably move it into a separate PR.

File: Makefile (hunk 5 above)
Comment (omgitsbillryan) [reply Makefile]: What was the original error? Perhaps there's a better solution. As it stands, setting `CI=true` & naming the make target `ci_local` is a bit of a confusing oxymoron - they mean the opposite of one another.

File: spec/models/institution_builder_spec.rb (hunk 6 above)
Comment (zurbergram) [reply spec/models/institution_builder_spec.rb]: https://github.com/department-of-veterans-affairs/gibct-data-service/pull/491

File: Makefile (hunk 5 above)
Comment (zurbergram) [reply Makefile]: Removed `ci_local`

File: spec/models/institution_builder_spec.rb (hunk 6 above)
Comment (zurbergram) [reply spec/models/institution_builder_spec.rb]: #491 is merged into master

**Review by kfrz — COMMENTED** [review]
- Neither [#492](https://github.com/department-of-veterans-affairs/gibct-data-service/pull/492) nor [#471](https://github.com/department-of-veterans-affairs/gibct-data-service/pull/492) are linked to issues on ZenHub - this makes it difficult for a reviewer to understand the context of the changes
- As far as I can I tell, there aren't robust user stories, AC, wireframes, or examples of what problem adding React specifically solves. This makes it hard to approve these changes. 
- I haven't seen any RFC, designs, informal discussion of potential alternatives. Do these exist?

> The idea behind adding React to this repo is that over time the UI can be written with React for easier maintainability. This is so that when switching from development in department-of-veterans-affairs/vets-website to development in department-of-veterans-affairs/gibct-data-service there is less of skill shift, 


- Adding frameworks increases surface area and also adds to maintenance requirements arguably reducing maintainability in the future.

> additionally there is functionality that React provides that without access to require "creative" solutions to be devised to achieve desired behavior in GIDS UI [src](https://github.com/department-of-veterans-affairs/gibct-data-service/pull/471#issuecomment-548850618)
- What particular pain points that aren't solvable with Rails + JS? [Rails + Turbolinks](https://guides.rubyonrails.org/working_with_javascript_in_rails.html) can accomplish perfectly performant implementations of most SPA-like behaviors.  
- Rails can render modern ES6 no problem and even utilize a full `webpack` based asset pipeline already, if there is a need for that - which doesn't require adding an additional framework

> The GIDS UI provides a tool for administrators to upload and manage data for the GI Bill Comparison Tool (GIBCT). The UI is only utilized by a couple of admin users, so performance is not really an issue. 

If the user base is only a few folks that's even more of a reason IMO to reduce overall SLOC and dependencies.

> There are several stories in the backlog to enhance the functionality of the GIDS UI to facilitate data maintenance. The team working the development stories also maintains the GIBCT utilizing React.

> Adding React to the GIDS UI will allow the team rapidly develop enhanced functionality for the GIDS UI working with familiar tools and not having to switch tool context when switching to GIDS UI. [src](https://github.com/department-of-veterans-affairs/gibct-data-service/pull/492#issuecomment-561269501)

I'd still like to see some more warrants for this change before providing explicit approval.
